### PR TITLE
freecad: 0.18.4 -> unstable-2020-09-25

### LINKED ISSUE #98821: Freecad 0.19
**Project description**
Add the freecad 0.19 package to the nixos repository to enable the mass exodus of fusion360 people to have the latest version

**Metadata**

* homepage URL: https://freecadweb.org/
* source URL: https://github.com/FreeCAD/FreeCAD
* license: mit, bsd, gpl2+ , ...GPLv2
* platforms: unix, linux, darwin, ... Windows, Linux, Mac

### PR DESCRIPTION
**Pull request by gebner**
###### Motivation for this change

We should be packaging the git version of FreeCAD instead of the latest "official" release.  AFAICT almost everybody else is using the nightlies anyhow.  [Upstream doesn't even accept bug reports for 0.18.4](https://tracker.freecadweb.org/my_view_page.php).

As a side effect, we can finally upgrade opencascade-occt and get rid of one more qtwebkit dependency.

Fixes #98821

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/graphics/freecad/default.nix [hunk 1]
```diff
@@ -9,13 +9,13 @@ let
   pythonPackages = python3Packages;
 in mkDerivation rec {
   pname = "freecad";
-  version = "0.18.4";
+  version = "unstable-2020-09-25";
```
Comment (mmilata) [inline pkgs/applications/graphics/freecad/default.nix:12]: Version must start with a digit, "unstable" should be part of pname: https://nixos.org/manual/nixpkgs/stable/#sec-package-naming

**Review by mmilata — COMMENTED** [review]
Things are sometimes broken in the nightlies, I wonder if it would make sense to offer both, e.g. as `freecad` and `freecad-stable` packages?

File: pkgs/applications/graphics/freecad/default.nix (hunk 1 above)
Comment (gebner) [reply pkgs/applications/graphics/freecad/default.nix]: Done.

**Comment by gebner** [comment]
> Things are sometimes broken in the nightlies, I wonder if it would make sense to offer both, e.g. as `freecad` and `freecad-stable` packages?

Ok, I've added back a `freecadStable` package.

**Comment by redvers** [comment]
Fair warning, you're likely you be asked to squash these commits.  The general rule is one commit per PR.

We may also want to have a conversation about directly tracking their master.

My concern is two-fold:
1. If a user includes freecad in their systemPackages and one day it ceases to compile due to say, a change in dependencies or their buildsystem then those users stuck with a system that won't build or, if they remove it - no way to install freecad until it is updated.
2. It's going to make caching near-useless.

I suggest instead of tracking master directly, specify a hash and version it with the date.

For example, the current would have a rev of 7616153b3c31ace006169cdc2fdafab484498858, and a version of "unstable-2020-09-25".

Does that make sense?

**Comment by gebner** [comment]
> Fair warning, you're likely you be asked to squash these commits. The general rule is one commit per PR.

I'm not sure where you got that from.  Usually we have one commit per changed package (with a commit message like `<package name>: 1.0 -> 2.0` or `<package name>: init at 1.0` or `<package name>: <other change>`).  Commits for multiple packages should *not* be squashed together.  Please also don't include merge commits in PRs if possible.

> We may also want to have a conversation about directly tracking their master. [...]

I think there is a slight misunderstanding here.  This PR updates FreeCAD to the latest master version *as of now*.  This version will not change until we update it the next time, in a separate PR.  It will not suddenly break if new commits are pushed to the FreeCAD repository.

**Review by mmilata — APPROVED** [review]
Didn't test but LGTM.

**Comment by redvers** [comment]
> > Fair warning, you're likely you be asked to squash these commits. The general rule is one commit per PR.
> 
> I'm not sure where you got that from. Usually we have one commit per changed package (with a commit message like `<package name>: 1.0 -> 2.0` or `<package name>: init at 1.0` or `<package name>: <other change>`). Commits for multiple packages should _not_ be squashed together. Please also don't include merge commits in PRs if possible.

Thank you for the correction.  I don't know how I misread that.  Maybe I shouldn't be looking at PRs before midday? :-)

> I think there is a slight misunderstanding here. This PR updates FreeCAD to the latest master version _as of now_. This version will not change until we update it the next time, in a separate PR. It will not suddenly break if new commits are pushed to the FreeCAD repository.

Perfect!

**Comment by jonringer** [comment]
I think it should have been `freecad` and `freecadUnstable` as that's similar to other already established paradigms like `zfs` and `zfsUnstable`

**Comment by gebner** [comment]
At first I only wanted to upgrade `freecad` (to the git version).  Personally, I would have only kept the current version.

If you think we should change the names to `freecad` and `freecadUnstable` now, we can also do that.

**Comment by FRidh** [comment]
Given upstream does not support their stable version, I think it is fine with the current naming. I would even say we should drop `freecadStable` for exactly that reason. Of course, if there are users that want the stable one we could choose to still provide it.

**Comment by jonringer** [comment]
> Given upstream does not support their stable version, I think it is fine with the current naming.

ah, I was unaware of this

**Pull request merged**